### PR TITLE
tests: check folder exist before creation

### PR DESCRIPTION
**Pull request by SagarGi**
## Description

This PR added code to the test to determine whether the file or folder existed before the folder's creation.

## Problem
https://github.com/nextcloud/integration_openproject/blob/master/tests/acceptance/features/api/setup.feature#L545
Previously, when a folder was created and it already existed, the server would return a 405 status code. But, now it is returning a 500 status code.

https://github.com/nextcloud/integration_openproject/blob/master/tests/acceptance/features/api/setup.feature#L679
Similarly, in case of the PROFIND, it is also returning a 500 status code.

Due to this reasons, there raise CI failure

```feature
    And user "OpenProject" has created folder "/OpenProject/project-abc"                                              # FeatureContext::userHasCreatedFolder()
      HTTP status code was not 201 while trying to create folder '/OpenProject' for user 'OpenProject'
      Failed asserting that an array contains 500.
```

> **_NOTE:_**  This problem raise on image juliushaertl/nextcloud-dev-php83:[20231202-1](https://github.com/users/juliushaertl/packages/container/nextcloud-dev-php83/153801076?tag=20231202-1) since on CI this image is used.

## Workpackage

- https://community.openproject.org/projects/nextcloud-integration/work_packages/57699

## Steps to reproduce
1. Run nextcloud using following command

```console
docker run --rm -p 8080:80 -e SERVER_BRANCH=stable30 -e NEXTCLOUD_AUTOINSTALL=YES -e NEXTCLOUD_AUTOINSTALL_APPS="groupfolders integration_openproject" -v ~/www/stable30/apps:/var/www/html/apps-extra/appid ghcr.io/juliushaertl/nextcloud-dev-php83:20231202-1
```

2.  Run failure test only

```console
NEXTCLOUD_BASE_URL=http://localhost/stable30 make api-test FEATURE_PATH=tests/acceptance/features/api/setup.feature:491
```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Tests only (no source changes)

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Updated `CHANGELOG.md` file

### CODE REVIEW

**Comment by nabim777** [comment]
https://github.com/nextcloud/integration_openproject/blob/master/tests/acceptance/features/api/setup.feature#L545
Previously, when a folder was created and it already existed, the server would return a 405 status code. But, now it is returning a 500 status code.

https://github.com/nextcloud/integration_openproject/blob/master/tests/acceptance/features/api/setup.feature#L679
Similarly, in case of the PROFIND, it is also returning a 500 status code.

@individual-it @SagarGi

**Comment by individual-it** [comment]
That sounds like a bug in NC, please 
1. have a look if there is already an issue in the server repo and open one if not
2. Adjust our test code, so that it can react the same regardless if the return is 405 or 500

**Comment by SagarGi** [comment]
> That sounds like a bug in NC, please
> 
> 1. have a look if there is already an issue in the server repo and open one if not
> 2. Adjust our test code, so that it can react the same regardless if the return is 405 or 500

I am quite confused that locally the NC seems to be giving correct status code. But using the docker image that we use in the CI for NC it gives 500. May be there seems to be something wrong with the image.

**Comment by nabim777** [comment]
Issue related to the PROFIND has been fixed in the image juliushaertl/nextcloud-dev-php83:[20231202-1](https://github.com/users/juliushaertl/packages/container/nextcloud-dev-php83/153801076?tag=20231202-1) 
but not in the folder creation

File: tests/acceptance/features/bootstrap/FeatureContext.php [hunk 1]
```diff
@@ -692,8 +696,8 @@ public function verifyTableNodeColumns(TableNode $table, ?array $requiredHeader
 		}
 	}
 
-	public function getIdOfElement(string $user, string $element): int {
-		$propfindResponse = $this->makeDavRequest(
+	public function getElementResponse(string $user, string $element): ResponseInterface {
```
Comment (SagarGi) [inline tests/acceptance/features/bootstrap/FeatureContext.php:699]:
```suggestion
	public function propfindFileOrFolder(string $user, string $path): ResponseInterface {
```

File: tests/acceptance/features/bootstrap/FeatureContext.php [hunk 2]
```diff
@@ -706,14 +710,23 @@ public function getIdOfElement(string $user, string $element): int {
 					  </d:prop>
 					</d:propfind>'
 		);
-		$xmlBody = $propfindResponse->getBody()->getContents();
-		$responseXmlObject = new SimpleXMLElement($xmlBody);
+	}
+
+	public function getIdOfElement(string $user, string $element): int {
+		$propfindResponse = $this->getElementResponse($user, $element);
```
Comment (SagarGi) [inline tests/acceptance/features/bootstrap/FeatureContext.php:716]:
```suggestion
		$propfindResponse = $this->propfindFileOrFolder($user, $path);
```

File: tests/acceptance/features/bootstrap/FeatureContext.php [hunk 3]
```diff
@@ -706,14 +710,23 @@ public function getIdOfElement(string $user, string $element): int {
 					  </d:prop>
 					</d:propfind>'
 		);
-		$xmlBody = $propfindResponse->getBody()->getContents();
-		$responseXmlObject = new SimpleXMLElement($xmlBody);
+	}
+
+	public function getIdOfElement(string $user, string $element): int {
```
Comment (SagarGi) [inline tests/acceptance/features/bootstrap/FeatureContext.php:715]:
```suggestion
	public function getIdOfFileOrFolder(string $user, string $path): int {
```

File: tests/acceptance/features/bootstrap/FeatureContext.php [hunk 4]
```diff
@@ -706,14 +710,23 @@ public function getIdOfElement(string $user, string $element): int {
...
+		$propfindResponse = $this->getElementResponse($user, $element);
+		$this->theHTTPStatusCodeShouldBe(207, "", $propfindResponse);
```
Comment (SagarGi) [inline tests/acceptance/features/bootstrap/FeatureContext.php:717]:
```suggestion
		$this->theHTTPStatusCodeShouldBe(207, "", $propfindResponse);
```
write a readable message for this?

File: tests/acceptance/features/bootstrap/FeatureContext.php [hunk 5]
```diff
@@ -706,14 +710,23 @@ public function getIdOfElement(string $user, string $element): int {
...
+		$responseXmlObject = new SimpleXMLElement($propfindResponse->getBody()->getContents());
 		$responseXmlObject->registerXPathNamespace(
 			'oc',
 			'http://owncloud.org/ns'
 		);
 		return (int)(string)$responseXmlObject->xpath('//oc:fileid')[0];
 	}
+
+	public function isElementExist(string $user, string $element): bool {
+		return $this->getElementResponse($user, $element)->getStatusCode() === 207;
```
Comment (SagarGi) [inline tests/acceptance/features/bootstrap/FeatureContext.php:727]:
```suggestion
		return $this->propfindFileOrFolder($user, $path)->getStatusCode() === 207;
```

File: tests/acceptance/features/bootstrap/FeatureContext.php [hunk 6]
```diff
@@ -706,14 +710,23 @@ public function getIdOfElement(string $user, string $element): int {
...
+		$this->theHTTPStatusCodeShouldBe(207, "", $propfindResponse);
+		$responseXmlObject = new SimpleXMLElement($propfindResponse->getBody()->getContents());
 		$responseXmlObject->registerXPathNamespace(
 			'oc',
 			'http://owncloud.org/ns'
 		);
 		return (int)(string)$responseXmlObject->xpath('//oc:fileid')[0];
 	}
+
+	public function isElementExist(string $user, string $element): bool {
```
Comment (SagarGi) [inline tests/acceptance/features/bootstrap/FeatureContext.php:726]:
```suggestion
	public function isElementExist(string $user, string $path): bool {
```

Comment (SagarGi) [inline tests/acceptance/features/bootstrap/FeatureContext.php:726]:
```suggestion
	public function fileOrFolderExists(string $user, string $path): bool {
```

File: tests/acceptance/features/bootstrap/FeatureContext.php [hunk 7]
```diff
@@ -345,14 +345,18 @@ public function userHasCreatedFolder(
 				continue;
 			}
 			$fullFolderString .= "/" . trim($folder);
+			// check if the file or folder already exists
+			if ($this->isElementExist($user, $fullFolderString)) {
```
Comment (SagarGi) [inline tests/acceptance/features/bootstrap/FeatureContext.php:349]:
```suggestion
			if ($this->fileOrFolderExists($user, $fullFolderString)) {
```

**Review by SagarGi — COMMENTED** [review]
LGTM :+1:

**Review by individual-it — APPROVED** [review]

**Pull request merged**